pkg/controller/jobframework/reconciler.go

```diff
@@ -376,15 +364,15 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, object clie
 		}
 	}
 
-	log.V(3).Info("restore podSetsInfo from annotation")
-	info, err := getPodSetsInfoFromObjectAnnotation(object, job)
-	if err != nil {
-		log.V(3).Error(err, "Unable to get original podSetsInfo")
-	} else {
+	if wl != nil {
+		log.V(3).Info("restore podSetsInfo")
+		info := getPodSetsInfoFromSpec(wl)
 		job.RestorePodSetsInfo(info)
-		return r.client.Update(ctx, object)
+		err := r.client.Update(ctx, object)
```

Comment:
> For now let's update the status first, if its API call fails , we are still suspended and will retry

The job controller will reset the value :(
https://github.com/kubernetes/kubernetes/blob/4706516100649bf5a92c780488ea53924239ad17/pkg/controller/job/job_controller.go#L767

It needs to be in the original order.

> If the update with restored pod-sets info fails, we should get to a point where the workload is not equivalent and the workload to be recreated, unfortunately with the wrong info.

Why would the equivalency check fail? Doesn't it account for the admission struct?
IIRC, we keep the admission struct until the Job is suspended.

> For the finalizes, even if we don't get to successfully do the update now, it will be done when the workload is deleted L155.

I rather not think about finalizers yet. Let's solve the possible race first.

Reply:
> Why would the equivalency check fail? Doesn't it account for the admission struct?
> IIRC, we keep the admission struct until the Job is suspended.

The job is suspended at that point.

> I rather not think about finalizers yet. Let's solve the possible race first.

Unless the job allows us to change the template and transition the suspended field at the same time, the sequence will always be problematic, there is no "right" way to manage a potential conflict here since the API changes (from outside kueue) may or may not, intentionally change the NodeSelector.

Reply:
Could we change the equivalency check so that it works even if the Job is suspended?

> Unless the job allows us to change the template and transition the suspended field at the same time

So today there are 3 steps: flip suspend, reset start time, change node selector. The reset start time is rather a hack from kueue, because the Job API doesn't actually support mutating node selectors when a job has already started.

The Job API does this to prevent a situation where you quickly flip suspend, change selectors and unsuspend. Then you could end up with running pods that have different templates.

I can see two possible futures:
1. The job controller automatically resets the start time to signal that it has successfully finished suspending a job. This would mean that kueue has to wait for the start time to reset before modifying the node selectors.
2. The job controller accepts mutations when suspending, but it has extra checks so that it doesn't create new pods if there are old pods that don't match the new template. There are possibilities for races here, but I think they are solvable. Perhaps we can pursue this in the next k8s release (1.29).

For now, can you check if the equivalency methods can be changed so that we can reset node selectors in case of failures? Again, I'm pretty sure conflicts are likely in a real cluster.

Reply:
Added  retry to restore the info and removing the finalizer.